pkg/model/iam.go

```diff
@@ -51,133 +51,158 @@ const RolePolicyTemplate = `{
-				},
-			}
+	}
 
-			// This is slightly tricky; we need to know the hosted zone id,
-			// but we might be creating the hosted zone dynamically.
+	{
+		iamPolicy := &iam.PolicyResource{
+			Builder: &iam.PolicyBuilder{
+				Cluster: b.Cluster,
+				Role:    igRole,
+				Region:  b.Region,
+			},
+		}
 
-			// TODO: I don't love this technique for finding the task by name & modifying it
-			dnsZoneTask, found := c.Tasks["DNSZone/"+b.NameForDNSZone()]
-			if found {
-				iamPolicy.DNSZone = dnsZoneTask.(*awstasks.DNSZone)
-			} else {
-				glog.V(2).Infof("Task %q not found; won't set route53 permissions in IAM", "DNSZone/"+b.NameForDNSZone())
-			}
+		// This is slightly tricky; we need to know the hosted zone id,
+		// but we might be creating the hosted zone dynamically.
 
-			t := &awstasks.IAMRolePolicy{
-				Name:      s(name),
-				Lifecycle: b.Lifecycle,
+		// TODO: I don't love this technique for finding the task by name & modifying it
+		dnsZoneTask, found := c.Tasks["DNSZone/"+b.NameForDNSZone()]
+		if found {
+			iamPolicy.DNSZone = dnsZoneTask.(*awstasks.DNSZone)
+		} else {
+			glog.V(2).Infof("Task %q not found; won't set route53 permissions in IAM", "DNSZone/"+b.NameForDNSZone())
+		}
 
-				Role:           iamRole,
-				PolicyDocument: iamPolicy,
-			}
-			c.AddTask(t)
+		t := &awstasks.IAMRolePolicy{
+			Name:      s(iamName),
+			Lifecycle: b.Lifecycle,
+
+			Role:           iamRole,
+			PolicyDocument: iamPolicy,
 		}
+		c.AddTask(t)
+	}
 
-		var iamInstanceProfile *awstasks.IAMInstanceProfile
-		{
-			iamInstanceProfile = &awstasks.IAMInstanceProfile{
-				Name:      s(name),
-				Lifecycle: b.Lifecycle,
-			}
-			c.AddTask(iamInstanceProfile)
+	var iamInstanceProfile *awstasks.IAMInstanceProfile
+	{
+		iamInstanceProfile = &awstasks.IAMInstanceProfile{
+			Name:      s(iamName),
+			Lifecycle: b.Lifecycle,
+			Shared:    fi.Bool(shared),
 		}
+		c.AddTask(iamInstanceProfile)
+	}
 
-		{
-			iamInstanceProfileRole := &awstasks.IAMInstanceProfileRole{
-				Name:      s(name),
-				Lifecycle: b.Lifecycle,
+	{
+		iamInstanceProfileRole := &awstasks.IAMInstanceProfileRole{
+			Name:      s(iamName),
+			Lifecycle: b.Lifecycle,
 
-				InstanceProfile: iamInstanceProfile,
-				Role:            iamRole,
-			}
-			c.AddTask(iamInstanceProfileRole)
+			InstanceProfile: iamInstanceProfile,
+			Role:            iamRole,
 		}
+		c.AddTask(iamInstanceProfileRole)
+	}
 
-		// Generate additional policies if needed, and attach to existing role
-		{
-			additionalPolicy := ""
-			if b.Cluster.Spec.AdditionalPolicies != nil {
-				roleAsString := reflect.ValueOf(role).String()
-				additionalPolicies := *(b.Cluster.Spec.AdditionalPolicies)
-				additionalPolicy = additionalPolicies[strings.ToLower(roleAsString)]
-			}
+	// Generate additional policies if needed, and attach to existing role
+	{
+		additionalPolicy := ""
+		if b.Cluster.Spec.AdditionalPolicies != nil {
+			roleAsString := reflect.ValueOf(igRole).String()
```

Comment:
I know this is not your code, so let's not fix it in this PR.  But using reflection for this looks ugly.  We can refactor later but I think we can just do `string(igRole)`.  Am I correct?  If so I will file an issue.